src/txmempool.h

```diff
@@ -736,6 +739,55 @@ class CTxMemPool
      *  removal.
      */
     void removeUnchecked(txiter entry, MemPoolRemovalReason reason) EXCLUSIVE_LOCKS_REQUIRED(cs);
+public:
+    /** EpochGuard: RAII-style guard for using epoch-based graph traversal algorithms.
+     *     When walking ancestors or descendants, we generally want to avoid
+     * visiting the same transactions twice. Some traversal algorithms use
+     * std::set (or setEntries) to deduplicate the transaction we visit.
+     * However, use of std::set is algorithmically undesirable because it both
+     * adds an asymptotic factor of O(log n) to traverals cost and triggers O(n)
+     * more dynamic memory allocations.
+     *     In many algorithms we can replace std::set with an internal mempool
+     * counter to track the time (or, "epoch") that we began a traversal, and
+     * check + update a per-transaction epoch for each transaction we look at to
+     * determine if that transaction has not yet been visited during the current
+     * traversal's epoch.
+     *     Algorithms using std::set can be replaced on a one by one basis.
+     * Both techniques are not fundamentally incomaptible across the codebase.
+     * Generally speaking, however, the remaining use of std::set for mempool
+     * traversal should be viewed as a TODO for replacement with an epoch based
+     * traversal, rather than a preference for std::set over epochs in that
+     * algorithm.
+     */
+    class EpochGuard {
+        const CTxMemPool& pool;
+        public:
+        EpochGuard(const CTxMemPool& in);
+        ~EpochGuard();
+    };
+    // N.B. GetFreshEpoch modifies mutable state via the EpochGuard construction
+    // (and later destruction)
+    EpochGuard GetFreshEpoch() const EXCLUSIVE_LOCKS_REQUIRED(cs);
+
+    /** visited marks a CTxMemPoolEntry as having been traversed
+     * during the lifetime of the most recently created EpochGuard
+     * and returns false if we are the first visitor, true otherwise.
+     *
+     * An EpochGuard must be held when visited is called or an assert will be
+     * triggered.
+     *
+     */
+    bool visited(txiter it) const EXCLUSIVE_LOCKS_REQUIRED(cs) {
+        assert(m_has_epoch_guard);
+        bool ret = it->m_epoch >= m_epoch;
+        it->m_epoch = std::max(it->m_epoch, m_epoch);
+        return ret;
+    }
+
+    bool visited(Optional<txiter> it) const EXCLUSIVE_LOCKS_REQUIRED(cs) {
```

Comment:
Correct; it's not needed yet. I'm indifferent on if it stays or goes. I figured it's better for it to live in the same commit as the introduction of it's non-optioned version, but unused functions aren't great to introduce.

Given that:
1) It will be introduced in later work, relatively soon
2) There's already an ack for this specific revision -- meaning both that this function has already been reviewed and doesn't require re-review in the future, but also that we won't make suhas have to re-ack this change.
3) There's nothing detectable *wrong* with it

I'd rather just keep it for now and focus on getting out the PR that depends on it sooner.

Reply:
If `Optional<txiter> it` has no value, returned value `visited() == true` seems semantically a bit weird. Could function naming be improved?

Reply:
It's not that weird of a decision once you look at the code that depends on it.

Essentially, if I code something like:
```c++
for (auto& it : mapTx) {
    auto it2 = FindIterByPrevout(it->GetTransaction()->vin[0].prevout);
    if (visited(it2)) continue;
}
```

this is kind of semantic, because if the FindByPrevout returns nullopt then we want to skip it because we've (in theory) already traversed everything we could want to traverse.


@sdaftuar and I spent like at least an hour trying to come up with a better semantic name than the predecessor to visited, and decided that visited was sufficiently semantic. But if you have a better suggestion, would be happy to consider it, we just couldn't come up with something that wasn't either awkward sounding or less semantic in other cases.

Reply:
`needs_processing(it2)` or similar could work, but `visited` seems pretty fine to me. (I would prefer it if it more clearly indicated that it's not idempotent/const; ie `bool a = visited(it); bool b = visited(it);` could result in `!a && b`, so maybe `first_visit()` or similar would be better? But still, it's already fine)

Reply:
Things like needs_processing aren't great because there can be some sort of implication that it2 needs processing after it's been marked, and doesn't convey the action. What it is a bit better for is conveying the optionality aspect (something that's an nullopt doesn't need processing).

The reason we liked visited as a name is that there is a hint that it's not idempotent, that visited suggests both a query (was it2 visited already?) and a statement (it2 has been visited now).

Ultimately a names just a name, and I don't think visited is horribly unclear. If visited isn't sufficient of a name, I'd opt for removing the optional version of this function (as previously noted it's unused) and just inlining the optional handling where it's used (although it wouldn't quite be the same because of the guard assertion checks).

(One of the other reasons for introducing this PR now rather than later by the way is that there are two or more separate places in the code presently that use Optional txiters, introducing this in a common PR means that those changes can proceed independently of one another.)